File: server/koboService.js
/**
 * KoboToolbox KPI API v2 Service Module.
 * Connects securely to KoboToolbox servers with Token authentication.
 */

/**
 * Normalizes server URL ensuring protocol and clean slashes.
 * @param {string} url 
 * @returns {string}
 */
export function normalizeServerUrl(url = 'https://kf.kobotoolbox.org') {
  let cleaned = url.trim();
  if (!cleaned.startsWith('http://') && !cleaned.startsWith('https://')) {
    cleaned = `https://${cleaned}`;
  }
  return cleaned.replace(/\/+$/, '');
}

/**
 * Fetches raw submission data from KoboToolbox API v2 endpoint.
 * @param {object} params 
 * @returns {Promise<object>}
 */
export async function fetchKoboSubmissions({
  serverUrl = 'https://kf.kobotoolbox.org',
  assetUid,
  token,
  limit = 100,
  offset = 0
}) {
  if (!assetUid) {
    throw new Error('El identificador del formulario (Asset UID) es requerido.');
  }

  if (!token) {
    throw new Error('El token de API de KoboToolbox (KOBO_API_TOKEN) es requerido.');
  }

  const baseUrl = normalizeServerUrl(serverUrl);
  const endpoint = `${baseUrl}/api/v2/assets/${assetUid}/data/?limit=${limit}&offset=${offset}`;

  try {
    const response = await fetch(endpoint, {
      method: 'GET',
      headers: {
        'Authorization': `Token ${token}`,
        'Accept': 'application/json'
      }
    });

    if (!response.ok) {
      const errorText = await response.text();
      let message = `Kobo API HTTP Error ${response.status}: ${response.statusText}`;

      if (response.status === 401) {
        message = 'Error de Autenticación 401: Token de KoboToolbox no válido o expirado.';
      } else if (response.status === 404) {
        message = `Formulario 404: No se encontró el Asset UID '${assetUid}' en el servidor '${baseUrl}'.`;
      }

      const err = new Error(message);
      err.status = response.status;
      err.details = errorText;
      throw err;
    }

    const data = await response.json();
    return data;
  } catch (error) {
    if (error.status) throw error;
    throw new Error(`Error de conexión con KoboToolbox (${baseUrl}): ${error.message}`);
  }
}

/**
 * Validates connection and asset existence with Kobo API v2.
 * @param {object} params 
 * @returns {Promise<{ success: boolean, name?: string, uid?: string, error?: string }>}
 */
export async function validateKoboConnection({ serverUrl, assetUid, token }) {
  if (!token || !assetUid) {
    return { success: false, error: 'Token y Asset UID son obligatorios' };
  }

  const baseUrl = normalizeServerUrl(serverUrl);
  const endpoint = `${baseUrl}/api/v2/assets/${assetUid}/`;

  try {
    const response = await fetch(endpoint, {
      method: 'GET',
      headers: {
        'Authorization': `Token ${token}`,
        'Accept': 'application/json'
      }
    });

    if (response.ok) {
      const asset = await response.json();
      return {
        success: true,
        name: asset.name || 'Formulario KoboToolbox',
        uid: asset.uid || assetUid
      };
    } else {
      return {
        success: false,
        error: `Error HTTP ${response.status}: Autenticación fallida o formulario no existe.`
      };
    }
  } catch (err) {
    return {
      success: false,
      error: `No se pudo conectar con el servidor Kobo (${baseUrl}): ${err.message}`
    };
  }
}

File: backend/kobo_service.py
"""
KoboToolbox API v2 Async Service module using httpx for FastAPI.
"""
import httpx
from typing import Dict, Any, Tuple

def normalize_server_url(url: str = "https://kf.kobotoolbox.org") -> str:
    cleaned = url.strip()
    if not cleaned.startswith("http://") and not cleaned.startswith("https://"):
        cleaned = f"https://{cleaned}"
    return cleaned.rstrip("/")

async def fetch_kobo_submissions(
    server_url: str,
    asset_uid: str,
    token: str,
    limit: int = 100,
    offset: int = 0
) -> Dict[str, Any]:
    """Async fetch raw submission data from KoboToolbox API v2."""
    if not asset_uid:
        raise ValueError("El identificador del formulario (Asset UID) es requerido.")
    if not token:
        raise ValueError("El token de API de KoboToolbox (KOBO_API_TOKEN) es requerido.")

    base_url = normalize_server_url(server_url)
    endpoint = f"{base_url}/api/v2/assets/{asset_uid}/data/?limit={limit}&offset={offset}"

    async with httpx.AsyncClient(timeout=15.0) as client:
        response = await client.get(
            endpoint,
            headers={
                "Authorization": f"Token {token}",
                "Accept": "application/json"
            }
        )

        if not response.is_success:
            err_msg = f"Kobo API HTTP Error {response.status_code}: {response.reason_phrase}"
            if response.status_code == 401:
                err_msg = "Error de Autenticación 401: Token de KoboToolbox no válido o expirado."
            elif response.status_code == 404:
                err_msg = f"Formulario 404: No se encontró el Asset UID '{asset_uid}' en el servidor '{base_url}'."
            raise RuntimeError(err_msg)

        return response.json()

async def validate_kobo_connection(
    server_url: str,
    asset_uid: str,
    token: str
) -> Tuple[bool, Dict[str, Any]]:
    """Validates connection and returns asset metadata."""
    if not token or not asset_uid:
        return False, {"error": "Token y Asset UID son requeridos."}

    base_url = normalize_server_url(server_url)
    endpoint = f"{base_url}/api/v2/assets/{asset_uid}/"

    try:
        async with httpx.AsyncClient(timeout=10.0) as client:
            response = await client.get(
                endpoint,
                headers={
                    "Authorization": f"Token {token}",
                    "Accept": "application/json"
                }
            )

            if response.is_success:
                asset = response.json()
                return True, {
                    "name": asset.get("name", "Formulario KoboToolbox"),
                    "uid": asset.get("uid", asset_uid)
                }
            else:
                return False, {"error": f"Error HTTP {response.status_code}: Autenticación fallida o formulario no existe."}
    except Exception as e:
        return False, {"error": f"No se pudo conectar con el servidor Kobo ({base_url}): {str(e)}"}
